### PR TITLE
Feature lwdev 8781 invalidate cache

### PR DESCRIPTION
**Pull request by OttoVT**

### CODE REVIEW

File: src/Lykke.HttpClientGenerator/Caching/ClientCacheManager.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+{
+    public class ClientCacheManager : IClientCacheManager
+    {
+        public ClientCacheManager()
+        {
+        }
+
+        public async Task InvalidateCacheAsync()
+        {
+            await (OnInvalidate?.Invoke() ?? Task.FromResult(0));
```
Comment (KonstantinRyazantsev) [inline src/Lykke.HttpClientGenerator/Caching/ClientCacheManager.cs]: Better to use Task.CompletedTask

File: src/Lykke.HttpClientGenerator/Caching/ClientCacheManager.cs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+        }
+
+        public event InvalidateCache OnInvalidate;
+
+        public void Dispose()
+        {
+            if (OnInvalidate != null)
```
Comment (KonstantinRyazantsev) [inline src/Lykke.HttpClientGenerator/Caching/ClientCacheManager.cs:20]: Usually subscribers are responsible to unsubscribe

File: src/Lykke.HttpClientGenerator/Caching/CachingCallsWrapper.cs [hunk 3]
```diff
@@ -16,32 +18,62 @@ namespace Lykke.HttpClientGenerator.Caching
...
             var contextData = new Dictionary<string, object>
             {
                 {ContextualTtl.TimeSpanKey, cachingTime}
             };
-            return _retryPolicy.ExecuteAsync((context, ct) => innerHandler(),
-                new Context(GetExecutionKey(targetMethod, args), contextData), default);
+
+            string cacheKey = GetExecutionKey(targetMethod, args);
+
+            using (var readLock = await _readerWriterLock.ReaderLockAsync())
```
Comment (KonstantinRyazantsev) [inline src/Lykke.HttpClientGenerator/Caching/CachingCallsWrapper.cs:55]: Inside this Read Lock, cache could be updated, so why this is read lock?

File: src/Lykke.HttpClientGenerator/Caching/ICustomAsyncCacheProvider.cs [hunk 4]
```diff
@@ -0,0 +1,14 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Polly.Caching;
+
+namespace Lykke.HttpClientGenerator.Caching
+{
+    public interface ICustomAsyncCacheProvider : IAsyncCacheProvider
```
Comment (KonstantinRyazantsev) [inline src/Lykke.HttpClientGenerator/Caching/ICustomAsyncCacheProvider.cs]: Let's nama it less general

File: src/Lykke.HttpClientGenerator/Caching/ClientCacheManager.cs (hunk 1 above)
Comment (OttoVT) [reply src/Lykke.HttpClientGenerator/Caching/ClientCacheManager.cs]: +

File: src/Lykke.HttpClientGenerator/Caching/CachingCallsWrapper.cs (hunk 3 above)
Comment (OttoVT) [reply src/Lykke.HttpClientGenerator/Caching/CachingCallsWrapper.cs:55]: In read lock we write cache keys in dictionary. In Invalidate method we simply clear cache+ keys in writeLock. It is being called often. While invalidate method is called rarely.

File: src/Lykke.HttpClientGenerator/Caching/ClientCacheManager.cs (hunk 2 above)
Comment (OttoVT) [reply src/Lykke.HttpClientGenerator/Caching/ClientCacheManager.cs:20]: In this situation the subscriber is a Call Wrapper which works like a middleware in calling pipeline.
And CacheManager is an external object to that pipeline, and only that object gives order to invalidate cache.

**Review by PaterSantyago — APPROVED** [review]

**Review by KonstantinRyazantsev — APPROVED** [review]

**Pull request merged**